packages/astro/src/vite-plugin-dev-toolbar/vite-plugin-dev-toolbar.ts

```diff
@@ -96,7 +120,62 @@ export default function astroDevToolbar({ settings, logger }: AstroPluginOptions
 						return undefined;
 					}
 				`;
-			}
+				}
+			},
+		},
+		{
+			name: 'astro:dev-toolbar:preact',
+			enforce: 'pre', // Need to be before Vite's plugins or they'll try to transform the JSX before us
+			resolveId(id) {
+				if (id === PUBLIC_VIRTUAL_MODULE_ID_PREACT) {
+					return preactResolvedVirtualModuleId;
+				}
+			},
+
+			async transform(code, id, options) {
+				const shouldTryTransform =
+					['.jsx', '.tsx'].includes(extname(removeQueryString(id))) && !options?.ssr;
+
+				if (!shouldTryTransform) {
+					return;
+				}
+
+				// TODO: Seems like a fairly old school way to detect JSX, maybe there's a better way?
+				const isJSXToolbarHeuristic =
+					code.includes('astro:toolbar:preact') ||
+					id.endsWith('?toolbar-app') ||
+					code.includes('astro/toolbar-preact');
```

Comment:
I tried a few ways before landing on this, most notably, bundling the entire toolbar app from its entrypoint. That worked, but it meant that apps had to be bundling compatible and on more complex apps, you instantly ran into warnings or errors (not to mention needing to fiddle with `external` and stuff for virtual modules etc)

This heuristic works just fine, however it is technically possible for a component to not trigger it:
- If it has no imports to `astro:toolbar:preact` (which means it doesn't use any features from Preact)
- Is not the entrypoint of the app
- Does not have a pragma comment

If anyone has a better solution, I'll take gladly. It's a bit tough to not conflict with Vite or worse, our renderers